[CI] CacheFileTests.testCacheFileCreatedAsSparseFile fails on master

**Build scan**:
https://gradle-enterprise.elastic.co/s/lny65mo6chmek

**Repro line**:
```
REPRODUCE WITH: ./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.xpack.searchablesnapshots.cache.common.CacheFileTests.testCacheFileCreatedAsSparseFile" \
  -Dtests.seed=E7AB31A9E717E54B \
  -Dtests.locale=zh-CN \
  -Dtests.timezone=Etc/GMT-5 \
  -Druntime.java=17

REPRODUCE WITH: ./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.xpack.searchablesnapshots.cache.common.CacheFileTests.testCacheFileCreatedAsSparseFile" \
  -Dtests.seed=E7AB31A9E717E54B \
  -Dtests.locale=zh-CN \
  -Dtests.timezone=Etc/GMT-5 \
  -Druntime.java=17
```

**Reproduces locally?**:
No

**Applicable branches**:
`master`

**Failure history**:
Two failures today.
No prior failures of this test.

**Failure excerpt**:
```
11:54:36 org.elasticsearch.xpack.searchablesnapshots.cache.common.CacheFileTests > testCacheFileCreatedAsSparseFile FAILED
11:54:36     java.lang.AssertionError
11:54:36         at __randomizedtesting.SeedInfo.seed([45B7968DCEE813EC:C0396B8309F7EF9D]:0)
11:54:36         at org.junit.Assert.fail(Assert.java:86)
11:54:36         at org.junit.Assert.assertTrue(Assert.java:41)
11:54:36         at org.junit.Assert.assertTrue(Assert.java:52)
11:54:36         at org.elasticsearch.xpack.searchablesnapshots.cache.common.CacheFileTests.testCacheFileCreatedAsSparseFile(CacheFileTests.java:422)
```

This test [failed again](https://gradle-enterprise.elastic.co/s/mm6twk4w5slek/tests/:x-pack:plugin:searchable-snapshots:test/org.elasticsearch.xpack.searchablesnapshots.cache.common.CacheFileTests/testCacheFileCreatedAsSparseFile?expanded-stacktrace=WyIwIl0&top-execution=1). It has a different error message but I guess it could still be related to this issue. Also since this issue is relatively new. I am re-openning it.

```
java.lang.AssertionError: Cache file size mismatches (block size: 8192, number of blocks: 97) |  
-- | --
  | Expected: <794624L> |  
  | but: was <798720L>
```

--- a/.ci/jobs.t/elastic+elasticsearch+periodic+ear.yml
+++ b/.ci/jobs.t/elastic+elasticsearch+periodic+ear.yml
@@ -25,6 +25,7 @@
           sudo cryptsetup open --key-file key.secret "$LOOP" secret --verbose
           sudo mkfs.ext2 /dev/mapper/secret
           sudo mkdir /mnt/secret
+          # Change /mnt/secret with care (at least a test makes assertion based on this name)
           sudo mount /dev/mapper/secret /mnt/secret
           sudo chown -R jenkins /mnt/secret
           cp -r "$WORKSPACE" /mnt/secret

--- a/server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java
+++ b/server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java
@@ -90,7 +90,7 @@ final class LinuxFileSystemNatives implements FileSystemNatives.Provider {
             final Stat stats = new Stat();
             final int rc = XStatLibrary.__xstat(STAT_VER, path.toString(), stats);
             if (logger.isTraceEnabled()) {
-                logger.trace("executing native method __xstat() returned {} with error code [{}] for file [{}]", rc, stats, path);
+                logger.trace("executing native method __xstat() returned {} with error code [{}] for file [{}]", stats, rc, path);
             }
             return OptionalLong.of(stats.st_blocks * ST_BLOCKS_UNIT);
         } catch (LastErrorException e) {

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
@@ -485,7 +485,7 @@ public class CacheFileTests extends ESTestCase {
                     assertThat(
                         "Non default block size only used in test executed with encryption at rest",
                         file.toAbsolutePath().toString(),
-                        startsWithIgnoringCase("/mnt/secret")
+                        containsString("/mnt/secret")
                     );
                 }
             }

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
@@ -55,7 +55,6 @@ import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
-import static org.hamcrest.Matchers.startsWithIgnoringCase;
 
 @LuceneTestCase.SuppressFileSystems("DisableFsyncFS") // required by {@link testCacheFileCreatedAsSparseFile()}
 public class CacheFileTests extends ESTestCase {

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
@@ -51,9 +51,11 @@ import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
+import static org.hamcrest.Matchers.startsWithIgnoringCase;
 
 @LuceneTestCase.SuppressFileSystems("DisableFsyncFS") // required by {@link testCacheFileCreatedAsSparseFile()}
 public class CacheFileTests extends ESTestCase {
@@ -436,13 +438,9 @@ public class CacheFileTests extends ESTestCase {
                 allOf(greaterThan(0L), lessThan(oneMb))
             );
 
-            final long blockSize;
-            if (Constants.LINUX) {
-                // on Linux we can infer the filesystem's block size if only 1 byte was written
-                blockSize = sizeOnDisk.getAsLong();
-            } else {
-                blockSize = 0L;
-            }
+            // on Linux we can infer the filesystem's block size if only 1 byte was written,
+            // but this is not always right with encryption at rest using dmcrypt
+            final long blockSize = Constants.LINUX ? sizeOnDisk.getAsLong() : 0L;
 
             fill(fileChannel, 0, Math.toIntExact(cacheFile.getLength()));
             fileChannel.force(false);
@@ -455,13 +453,41 @@ public class CacheFileTests extends ESTestCase {
                 greaterThanOrEqualTo(cacheFile.getLength())
             );
 
-            if (Constants.LINUX) {
+            if (blockSize > 0L) {
                 final long nbBlocks = (cacheFile.getLength() + blockSize - 1) / blockSize; // ceil(cacheFile.getLength() / blockSize)
-                assertThat(
-                    "Cache file size mismatches (block size: " + blockSize + ", number of blocks: " + nbBlocks + ')',
-                    sizeOnDisk.getAsLong(),
-                    equalTo(nbBlocks * blockSize)
-                );
+                final long expectedSize = nbBlocks * blockSize;
+                if (blockSize == fourKb) {
+                    assertThat(
+                        "Cache file size mismatches (block size: "
+                            + blockSize
+                            + ", number of blocks: "
+                            + nbBlocks
+                            + ", file length: "
+                            + cacheFile.getLength()
+                            + ')',
+                        sizeOnDisk.getAsLong(),
+                        equalTo(expectedSize)
+                    );
+                } else {
+                    // block size other than usual default block size of 4KB indicates that a special filesystem may be at use; in this
+                    // case we verify that the allocated size for the file stays within some upper limit of an extra block of 4096L bytes
+                    assertThat(
+                        "Cache file size mismatches with non default block size (block size: "
+                            + blockSize
+                            + ", number of blocks: "
+                            + nbBlocks
+                            + ", file length: "
+                            + cacheFile.getLength()
+                            + ')',
+                        sizeOnDisk.getAsLong(),
+                        allOf(greaterThanOrEqualTo(expectedSize), lessThanOrEqualTo(expectedSize + fourKb))
+                    );
+                    assertThat(
+                        "Non default block size only used in test executed with encryption at rest",
+                        file.toAbsolutePath().toString(),
+                        startsWithIgnoringCase("/mnt/secret")
+                    );
+                }
             }
         } finally {
             cacheFile.release(listener);

--- a/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
+++ b/x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/common/CacheFileTests.java
@@ -51,7 +51,6 @@ import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
-import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 import static org.hamcrest.Matchers.sameInstance;
@@ -468,19 +467,7 @@ public class CacheFileTests extends ESTestCase {
                         equalTo(expectedSize)
                     );
                 } else {
-                    // block size other than usual default block size of 4KB indicates that a special filesystem may be at use; in this
-                    // case we verify that the allocated size for the file stays within some upper limit of an extra block of 4096L bytes
-                    assertThat(
-                        "Cache file size mismatches with non default block size (block size: "
-                            + blockSize
-                            + ", number of blocks: "
-                            + nbBlocks
-                            + ", file length: "
-                            + cacheFile.getLength()
-                            + ')',
-                        sizeOnDisk.getAsLong(),
-                        allOf(greaterThanOrEqualTo(expectedSize), lessThanOrEqualTo(expectedSize + fourKb))
-                    );
+                    // block size other than usual default block size indicates that a special filesystem may be at use, let's verify this
                     assertThat(
                         "Non default block size only used in test executed with encryption at rest",
                         file.toAbsolutePath().toString(),

